rule/string-regex.go

```diff
@@ -0,0 +1,233 @@
+	if err != nil {
+		panic(fmt.Sprintf("unable to compile %s as regexp (argument %d, option 1)", rule[1], ruleNum))
+	}
+
+	// Use custom error message if provided
+	if len(rule) == 3 {
+		errorMessage = rule[2]
+	}
+	return scope, regex, errorMessage
+}
+
+// #endregion
+
+// #region Node traversal
+
+func (w lintStringRegexRule) Visit(node ast.Node) ast.Visitor {
+	// First, check if node is a call expression
+	call, ok := node.(*ast.CallExpr)
+	if !ok {
+		return w
+	}
+
+	// Get the name of the call expression to check against rule scope
+	callName, ok := w.getCallName(call)
+	if !ok {
+		return w
+	}
+
+	for _, rule := range w.rules {
+		if rule.Scope.Func == callName {
+			rule.Apply(call)
+		}
+	}
+
+	return w
+}
+
+// Return the name of a call expression in the form of package.Func or Func
+func (w lintStringRegexRule) getCallName(call *ast.CallExpr) (callName string, ok bool) {
+	if ident, ok := call.Fun.(*ast.Ident); ok {
+		// Local function call
+		return ident.Name, true
+	} else if selector, ok := call.Fun.(*ast.SelectorExpr); ok {
+		// Scoped function call
+		scope, ok := selector.X.(*ast.Ident)
+		if !ok {
+			return "", false
+		}
+		return scope.Name + "." + selector.Sel.Name, true
+	} else {
+		return "", false
+	}
+}
+
+// #endregion
+
+// #region Linting logic
+
+// Apply a single regex rule to a call expression (should be done after verifying the that the call expression matches the rule's scope)
+func (rule stringRegexSubrule) Apply(call *ast.CallExpr) {
+	if len(call.Args) <= rule.Scope.Argument {
+		// TODO: should cases where calls are incompatible with scope cause failures?
+		return
+	}
+
+	arg := call.Args[rule.Scope.Argument]
+	var lit *ast.BasicLit
+	if len(rule.Scope.Field) > 0 {
+		// Try finding the scope's Field, treating arg as a composite literal
+		composite, ok := arg.(*ast.CompositeLit)
+		if !ok {
+			return
+		}
+		for _, el := range composite.Elts {
+			kv, ok := el.(*ast.KeyValueExpr)
+			if !ok {
+				continue
+			}
+			key, ok := kv.Key.(*ast.Ident)
+			if !ok || key.Name != rule.Scope.Field {
+				continue
+			}
+
+			// We're now dealing with the exact field in the rule's scope, so if anything fails, we can safely return instead of continuing the loop
+			lit, ok = kv.Value.(*ast.BasicLit)
+			if !ok || lit.Kind != token.STRING {
+				return
+			}
+		}
+	} else {
+		var ok bool
+		// Treat arg as a string literal
+		lit, ok = arg.(*ast.BasicLit)
+		if !ok || lit.Kind != token.STRING {
+			return
+		}
+	}
+	// Unquote the string literal before linting
+	unquoted := lit.Value[1 : len(lit.Value)-1]
+	rule.lintMessage(unquoted, lit)
+}
+
+func (rule stringRegexSubrule) lintMessage(s string, node ast.Node) {
+	// Fail if the string doesn't match the user's regex
+	if rule.Regexp.MatchString(s) {
+		return
+	}
+	var failure string
+	if len(rule.ErrorMessage) > 0 {
+		failure = fmt.Sprintf("string literal doesn't match user defined regex (%s)", rule.ErrorMessage)
```

Comment:
why not just report the rule error message?
i.e. in place of
`string literal doesn't match user defined regex (must start with a capital letter)`
we could just emit
`must start with a capital letter`

up to the user to provide a nice error message